### PR TITLE
Add back cmake command-line cache on unix

### PR DESCRIPTION
**Pull request by sbomer**
This was deleted in https://github.com/dotnet/runtime/pull/64370 because a canceled configure would leave around the cache file, causing subsequent invocations to start building with the incomplete configure.

I'm adding it back along with a fix to write the cache file _after_ configure, as suggested in https://github.com/dotnet/runtime/pull/64370#issuecomment-1026994349. I made the same fix to the windows logic.

This saves 1-2s during incremental builds on my machine. Contributes to https://github.com/dotnet/runtime/issues/47022.

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

**Comment by janvorli** [comment]
Besides the interrupted build, there was another case that was making this caching painful for me. That was when I was adding or modifying certain feature detection in configure.cmake stuff. Such changes require the cmake to be rerun, so the caching was preventing these changes from having any effect and I had to manually delete the artifacts/obj directory each time I've made a change into that file. Actually, editing anything in the cmake files has no effect unless you delete the artifacts/obj if the caching is added. That might be unexpected to people when they update their local state from github without cleaning the artifacts folder.
Due to the issues I've mentioned, it doesn't seem worth the 1-2 secs gain in build time for me, but if we really want to make this change, then I'd like to at least have a command line option to disable this caching.

**Comment by am11** [comment]
This is saving a second for one use-case and adding 15 minutes penalty for others where there is no way but to delete the  `artifacts` directory and rebutild. cmake has its own caching mechanism which everyone uses and should come natural. I don't think this is the right place to save 1-2 seconds.

You can save time by deduplicating and deleting unused / unnecessary introspection checks. Plenty to clean up on the real side of things..

**Comment by vargaz** [comment]
I agree as well that the default build should be safe, and if somebody wants a shortcut build which is faster but can sometime fail, then there should be some kind of option to do that.

**Comment by janvorli** [comment]
After thinking about it a bit more, I feel more strongly now that the default should be to not to use the cache. We can add a command line option to enable the caching for the cases where we care about the extra second or two.
The reason is a danger that having the cache can cause people to get incorrect builds after syncing from github / making changes to the cmake build files. That is quite non-obvious, I would expect the whole build to be incremental including the cmake file changes. I really dislike cases when I make a change, it doesn't seem to work, so I have to spend time debugging and investigating what's wrong and then figuring out that it was actually a problem of the build not picking my changes. Or even worse, creating a build for testing at a customer that ends up not having intended fixes.

**Comment by janvorli** [comment]
@sbomer I have thought that your change was skipping cmake completely when the command line was detected to be unchanged. That would prevent the cmake logic you've mentioned from kicking in and that's exactly the concern I had. Am I missing something then?

**Comment by vargaz** [comment]
The generated makefiles/ninja files check the cmake files for updates. It can pick up changes to the files themselves, but not environment changes, etc.

**Comment by am11** [comment]
> I think that re-running cmake configure also doesn't pick up _environment_ changes so adding the cache file should have the same behavior.

It does. If you install a missing dependency and rerun the build command, it works with what we have in `main` but not with this hand-rolled cache.

**Comment by akoeplinger** [comment]
@am11 that's a different scenario, in that case the configure step fails before we write the cache file so it'd be run again the next time.

**Comment by am11** [comment]
> that's a different scenario,

This PR is reverting my PR which was fixing the exact scenario. Also see https://github.com/dotnet/runtime/issues/65533 for other issues with this kind of custom caching.

File: eng/Versions.props [hunk 1]
```diff
@@ -91,7 +91,7 @@
     <MicrosoftDotNetXUnitConsoleRunnerVersion>2.5.1-beta.23309.8</MicrosoftDotNetXUnitConsoleRunnerVersion>
     <MicrosoftDotNetBuildTasksArchivesVersion>8.0.0-beta.23309.8</MicrosoftDotNetBuildTasksArchivesVersion>
     <MicrosoftDotNetBuildTasksInstallersVersion>8.0.0-beta.23309.8</MicrosoftDotNetBuildTasksInstallersVersion>
-    <MicrosoftDotNetBuildTasksPackagingVersion>8.0.0-beta.23309.8</MicrosoftDotNetBuildTasksPackagingVersion>
+    <MicrosoftDotNetBuildTasksPackagingVersion>8.0.0-dev</MicrosoftDotNetBuildTasksPackagingVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:94]: Is this correct? Or was it for local testing?

**Comment by stephentoub** [comment]
@sbomer, this hasn't been touched in a year. Are you still working on it? Should it be closed?

**Pull request closed without merging**